### ISSUE TITLE
Compile top-level await to ES5

### ISSUE THREAD

**Issue by amannn** [webpack/webpack#11874]
# Bug report

**What is the current behavior?**

Top-level await code is currently always compiled to native async/await code and therefore doesn't work in browsers like IE11.

**If the current behavior is a bug, please provide the steps to reproduce.**

Reproduction repository: https://github.com/amannn/webpack-top-level-await

Here's the relevant part of [the output](https://github.com/amannn/webpack-top-level-await/blob/main/dist/main.js):

```js
!(function () {
  "use strict";
  var r = {
      568: function (r, o) {
        r.exports = (async () => ((o.Z = await Promise.resolve(1)), o))();
      },
      138: function (r, o, e) {
        r.exports = (async () => {
          var r = e(568);
          (r = await Promise.resolve(r)), console.log(r.Z);
        })();
      },
    },
    o = {};
  !(function e(t) {
    if (o[t]) return o[t].exports;
    var s = (o[t] = { exports: {} });
    return r[t](s, s.exports, e), s.exports;
  })(138);
})();
```

**What is the expected behavior?**

It would be great if the code would compile to regular promises that are awaited by calling `.then`.

I tried configuring `target: ['web', 'es5']` and adding a `browserslist` config, but it seems like that doesn't have an effect.

**Other relevant information:**
webpack version:  5.3.1
Node.js version: 14.15.0
Operating System: macOS
Additional tools: I initially observed this behaviour in a Next.js app and have created [a failing test in the Next.js repository](https://github.com/vercel/next.js/pull/18487).

**Comment by evilebottnawi** [comment]
It is invalid syntax, you can't use `await` in `export`

**Comment by evilebottnawi** [comment]
You need babel-loader to transform it for IE11

**Comment by evilebottnawi** [comment]
As you can see there is not magic for `topLevelAwait` https://github.com/webpack/webpack/blob/master/lib/dependencies/HarmonyDetectionParserPlugin.js#L51

**Comment by amannn** [comment]
Thanks for your quick response!

I've created a `babel-loader` branch based on your feedback in my repo: https://github.com/amannn/webpack-top-level-await/tree/babel-loader

> It is invalid syntax, you can't use await in export

Is it? It seems like it at least compiled. I've now changed the code to this:

```js
const value = await Promise.resolve(1);

export default value;
```

Does that look ok to you?

> You need babel-loader to transform it for IE11

I've added babel loader, `@babel/preset-env` and `@babel/plugin-transform-async-to-generator` in the branch now. I've also included some regular async/await code in the example code. The regular async/await code is transformed as expected, but not the top-level async/await.

I've read at https://babeljs.io/docs/en/babel-plugin-syntax-top-level-await that this plugin only parses the syntax but doesn't do any transformation itself.

As far as I understand, Babel runs before the webpack compilation, right? So in this case webpack inserts new async/await statements in the code that was already compiled by Babel, so Babel wouldn't be able to transform the result – is that correct?

**Comment by evilebottnawi** [comment]
Yep, you are right

**Comment by evilebottnawi** [comment]
Feel free to send a PR https://github.com/webpack/webpack/blob/master/lib/async-modules/AwaitDependenciesInitFragment.js

**Comment by amannn** [comment]
Thanks for the pointer! I've created an initial PR with a failing test (https://github.com/webpack/webpack/pull/11881), but I'm struggling a bit to understand how I can wrap the whole module consumption in a `.then(…)`. If you could help with some more guidance, that would be great! I'm a first-time contributor and not really familiar with the internals of webpack.

**Comment by evilebottnawi** [comment]
Just use `then` :smile:

**Comment by amannn** [comment]
I'm really not sure how 🙂. It looks like with `AwaitDependenciesInitFragment` I can add a line before the module exports are consumed, but how can I wrap the whole consumption? I'd need to insert code at the end as well, right?

**Comment by evilebottnawi** [comment]
try experimenting, just open `node_modules/webpack/lib/async-modules/AwaitDependenciesInitFragment.js` in your reproducible test repo and change runtime and look what webpack generated

**Comment by Janpot** [comment]
I took a look at this yesterday, and realised something. In order for untranspiled top-level await user code to work, webpack needs to provide it with an async context. The only way to do that is by [wrapping it with an `async` function](https://github.com/webpack/webpack/blob/d770a55ec0403fdf52fbd0fbeb8829ea3f28e574/lib/dependencies/HarmonyCompatibilityDependency.js#L75). We can't replace this `async` function with `.then` because that would remove the async context and make untranspiled top-level await fail. I think that in order to make top-level await compatible with ie11, webpack output will need to be transpiled in a post-processing step.

**Comment by alexander-akait** [comment]
bump

**Comment by alexander-akait** [comment]
Let's keep open, I think we can improve warnings/docs here

**Comment by omgoshjosh** [comment]
lol bad bot!

**Comment by vanceeasleaf** [comment]
I need this PR

**Comment by vanceeasleaf** [comment]
well , I write a wepack plugin to transform async to es5 
`class BabelTransformAsyncPlugin {
  apply(compiler) {
    compiler.hooks.compilation.tap(
      "BabelTransformAsyncPlugin",
      (compilation) => {
        compilation.hooks.optimizeChunkAssets.tapAsync(
          "BabelTransformAsyncPlugin",
          (chunks, done) => {
            chunks.forEach(function(chunk) {
              chunk.files.forEach(function(fileName) {
                var result = compilation.assets[fileName].source();
                console.log(fileName);
                if (fileName.endsWith(".js") && result.includes("async ")) {
                  console.log("babel async trans", fileName);

                  result = require("@babel/core").transformSync(result, {
                    configFile: false,
                    plugins: ["@babel/plugin-transform-async-to-generator"],
                  }).code;
                  // ...require("./babel.config"),

                  compilation.assets[fileName] = new ConcatSource(result);
                }
              });
            });
            done();
          }
        );
      }
    );
  }
}`

**Comment by siawo** [comment]
Hi @vanceeasleaf, Have you created the webapck plugin? If not how are you achieving the the top level await for IE11?

**Comment by vankop** [comment]
we discussed this internally so.. supporting promises on webpack side would require huge code transformation, that webpack in current version can not perform. There is a [PR](https://github.com/webpack/webpack/pull/15612) to add warning in case when output environment does not supports async/await.

As user you can:

- build for environment that supports async/await
- don't use top-level-await

**Comment by williamjacobszoon** [comment]
@vankop Just wanted to make sure that this issue will be fixed in a future version of webpack right? adding a warning is a temporary solution I guess.

**Comment by vankop** [comment]
we didn't plan this yet.

**Comment by alioshr** [comment]
> well , I write a wepack plugin to transform async to es5
> 
> ```js
> class BabelTransformAsyncPlugin {
>   apply(compiler) {
>     compiler.hooks.compilation.tap(
>       "BabelTransformAsyncPlugin",
>       (compilation) => {
>         compilation.hooks.optimizeChunkAssets.tapAsync(
>           "BabelTransformAsyncPlugin",
>           (chunks, done) => {
>             chunks.forEach(function(chunk) {
>               chunk.files.forEach(function(fileName) {
>                 var result = compilation.assets[fileName].source();
>                 console.log(fileName);
>                 if (fileName.endsWith(".js") && result.includes("async ")) {
>                   console.log("babel async trans", fileName);
> 
>                   result = require("@babel/core").transformSync(result, {
>                     configFile: false,
>                     plugins: ["@babel/plugin-transform-async-to-generator"],
>                   }).code;
>                   // ...require("./babel.config"),
> 
>                   compilation.assets[fileName] = new ConcatSource(result);
>                 }
>               });
>             });
>             done();
>           }
>         );
>       }
>     );
>   }
> }
> ```

Did you succeed with this plug-in? Thanks for sharing

**Comment by steverep** [comment]
This should probably remain open

**Comment by steverep** [comment]
So I decided to write a plugin as well for this and put it on NPM: https://www.npmjs.com/package/transform-async-modules-webpack-plugin.

It's kind of similar to the one pasted above by @vanceeasleaf in that it uses Babel, but with some key advantages:

- Handles source maps
- Operates on the modules directly before they are written to chunks instead of transforming the chunks, meaning  Babel helper injections won't pollute the global scope and it still works if an "eval" `devTool` is used
- Uses `@babel/preset-env` to be able to just specify targets and so that the generators can also be transpiled down to ES5 (just `plugin-transform-async-to-generator` is not enough)

Functionally it seems to work fine.  The biggest issue I have is that it doesn't use `@babel/plugin-transform-runtime` because it's way too late in the Webpack process to handle injected imports.  So if you have a lot of TLAs or even just 1 or 2 that permeate and result in a lot of downstream async modules, then the repeated helper and regenerator functions add up quickly in a large bundle.  I plan to try and address this: https://github.com/steverep/transform-async-modules-webpack-plugin/issues/1.

@alexander-akait  or @vankop if you have any ideas or advice on the best way to seamlessly add the `@babel/runtime` dependencies, I would appreciate it.  I'm looking at doing something similar to the way `ProvidePlugin` works, but I'm just starting to get myself familiar with how the `InitFragment` class works.  I'm not sure if there is a hook order that will work to replace the imports or if I need to flag and rebuild to make it work.

**Comment by vankop** [comment]
at least as I know there is no plan to support code transformation since this will require huge internal webpack changes. old `Dependency` design will no longer work since dependencies operate not on AST level of source code, but only insert/remove/change strings inside source code.

Unfortunately, I am not familiar with babel, but I guess transforming top level await does not help.. since it is required to `await` module import. e.g. in example:


module A
```
const a = await somePromise;
export { a }
```

module B
```
import { a } from 'module-a';
// here we should wait for promise 'somePromise'

console.log(a);
```

in case of transformation just async/await this will not work correctly since you need to transform hole module tree ( thats an issue with currect webpack design, we dont have all code as an AST internally )

To sum up. as a user you can do things that [I mentions above](https://github.com/webpack/webpack/issues/11874#issuecomment-1100652054) + chunk transformation ( this will bring additional code unfortunately )

**Comment by steverep** [comment]
@vankop I think maybe you are misunderstanding how the plugin and transformations work.  I don't think huge Webpack changes are necessary to support this.

> Unfortunately, I am not familiar with babel, but I guess transforming top level await does not help.

As you probably know, Babel cannot transform top level await; it simply supports the syntax to possibly transpile the expression, but otherwise leaves the `await` as is.  However, once the `await` is inside an `async` function,it transpiles that function by:
1. Turning it into a generator function that yields a promise
2. Turning that generator into a regular function using a state machine that returns the promise

> in case of transformation just async/await this will not work correctly since you need to transform hole module tree 

That's not necessary. Once Webpack wraps the original module in an `async function`, the new modified module can be transformed by Babel with no other knowledge of the tree.  My plugin does this by tapping into the `renderModuleContent` hook in the `JavascriptModulesPlugin`, so right before it gets wrapped in a container for the chunk.  This works fine.

The enhancement I'd like to make is regarding helper code.  As part of that transformation, Babel inserts some expensive helpers at the top of the module:

```js
var helper1 = function (...) // expensive state machine function
var helper2 = function (...) // expensive asyncToGenerator converter function
... // Transpiled version of async-wrapped module
```

Instead of repeating those expensive helpers over and over again, Babel can instead import them from  the `@babel/runtime` package:

```js
var helper1 = require("@babel/runtime/helper1");
var helper2 = require("@babel/runtime/helper2");
... // Transpiled version of async-wrapped module
```

So my question is: is there a way I can not repeat those helpers by replacing those requires with something Webpack understands?  From what I can see, there's no way to add dependencies at that stage as the graphs are already built and optimized.  And there's no way to know which modules are wrapped in async functions without building and optimizing the graph first.  Correct me if I'm wrong there.


My only thought was to flag the async modules on 1st pass, then add the dependencies on a call to `module.rebuild()`, but that seems risky as the graph could keep changing?  Or maybe the helpers could be added to Webpack's runtime code somehow to avoid all that?

**Comment by vankop** [comment]
you need to do this on loader level. `modulGraph.isAsync` returns `true` for any module that imports top-level-await module.

**Comment by steverep** [comment]
> you need to do this on loader level. `modulGraph.isAsync` returns `true` for any module that imports top-level-await module.

I don't see how to use the loader interface to accomplish it at all.  You don't know where the TLAs are without parsing and you don't know which others import it until the whole graph is finished, right?

These imports also cannot be treated like any other dependency because then they'll be wrapped inside the async function.  They need to be in the module scope.

**Comment by alexander-akait** [comment]
Let's keep open if sombody wants to improve it, right now I think we need to output a warning for our users to prevent some problems on production (when you have lower the `target` option)

**Comment by steverep** [comment]
For anyone interested, I did make the [plugin I wrote](https://github.com/steverep/transform-async-modules-webpack-plugin) work with `@babel/runtime` pretty seamlessly, so I consider it pretty much a finished product that I'll make an effort to maintain.

@alexander-akait I'm not sure what criteria you'd hope to achieve to close this, but I could certainly offer a PR to point the docs and/or error message to my plugin.  Alternatively, I could try putting it directly in Webpack, but that would be a big chore for me and introduce a dependency on Babel that I'm not sure would be desired.

IMO, a solution here without using an external transpiler certainly has advantages, but in the end reinvents the wheel and becomes unnecessarily burdensome for Webpack maintainers.

**Comment by alexander-akait** [comment]
@steverep Thank you for your work and link, let's put it in our docs on site :+1:
